sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -956,9 +956,19 @@ object TypeCoercion extends TypeCoercionBase {
       })
   }
 
+  @tailrec
   override def implicitCast(e: Expression, expectedType: AbstractDataType): Option[Expression] = {
-    implicitCast(e.dataType, expectedType).map { dt =>
```

Comment:
Sorry I didn't make it clear. I mean the inner `implicitCast`, which has two implementations